### PR TITLE
Makes it easier to use custom Routes within WorkboxSW.

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @addyosmani @gauntface
CC: @jmtt89 @mikefowler

This is something of a straw-man PR, and feedback is welcome.

This partially addresses #385 and #636, but I'm not sure that I'd call it a "fix", so I'm inclined to leave them open.

Still, it should make it easier to accomplish what @jmtt89 was trying to do, by letting you use a method other than `'GET'` when calling `workboxSW.router.registerRoute()`.

It also provides an escape hatch granting greater flexibility by letting developers import and use `Route` objects created from the `workbox.routing` namespace, by relaxing the assertions to just check whether the object passed to `registerRoute()` has the properties that a `Route` should have, not necessarily whether it's a subclass of a specific definition of a `Route` object.

This PR doesn't attempt to add in a factory method for creating new `Route` objects exposed via`workboxSW.router`, although that's an idea that's been discussed in the past and might still be on the table.

### CODE REVIEW

**Comment by mikefowler** [comment]
@jeffposnick this seems like a reasonable solution to me. 

What was the original/current thinking on not including `Route` in the public interface for an instance of `WorkboxSW`? I'll admit, it was confusing to see the class in the source code but have no way of accessing it.

**Comment by jeffposnick** [comment]
I don't remember the original thinking—@gauntface and I had some back and forth about it, but I'm not even sure which side each of us took 😄 

It's slightly awkward to expose the `Route` class directly, because `workboxSW.router` is an automatically-constructed instance of a subclass of `Router`.

We could expose a `workboxSW.router.createRoute()` factory method (though that would preclude subclassing), or we could expose the `Route` class statically as `workboxSW.Route`.

I guess my vote is for the factory method, but that can be implemented separately from this PR.

File: lib/assert.js [hunk 1]
```diff
@@ -91,6 +91,36 @@ export function isArrayOfClass(object, expectedClass) {
   }
 }
 
+export function isInterface(object, expectedProperties) {
+  const parameter = Object.keys(object).pop();
+  const message = `The '${parameter}' parameter should be an object that has
```
Comment (gauntface) [inline lib/assert.js]: Should you be creating this inside the for loop to update the message for the failing property?

File: lib/assert.js [hunk 2]
```diff
@@ -91,6 +91,36 @@ export function isArrayOfClass(object, expectedClass) {
...
+    }
+  }
+}
+
+export function isArrayOfInterface(object, expectedProperties) {
+  const parameter = Object.keys(object).pop();
+  const message = `The '${parameter}' parameter should be an array containing
+    objects that each have '${expectedProperties}' properties.`;
+
+  if (!Array.isArray(object[parameter])) {
```
Comment (gauntface) [inline lib/assert.js]: not 100% of this check, the function name 'isArrayOfInterface' suggests that object should be an array and this asserts each entry in the array has an explicit interface.

File: lib/assert.js [hunk 3]
```diff
@@ -91,6 +91,36 @@ export function isArrayOfClass(object, expectedClass) {
...
+    throwError(message);
+  }
+
+  for (let item of object[parameter]) {
+    for (let property of expectedProperties) {
+      if (!(property in item)) {
+        throwError(message);
```
Comment (gauntface) [inline lib/assert.js]: same as above, the message should be updated to handle the missing property

File: packages/workbox-routing/src/lib/router.js [hunk 4]
```diff
@@ -13,11 +13,12 @@
  limitations under the License.
 */
 
-import Route from './route';
-import {isArrayOfClass, isInstance} from '../../../../lib/assert';
+import {isArrayOfInterface, isInterface} from '../../../../lib/assert';
 import logHelper from '../../../../lib/log-helper.js';
 import normalizeHandler from './normalize-handler';
 
+const ROUTE_INTERFACE = ['match', 'handler', 'method'];
```
Comment (gauntface) [inline packages/workbox-routing/src/lib/router.js]: I'm not a huge fan of having this manually defined and it being defined away from the class it represents.

File: lib/assert.js [hunk 5]
```diff
@@ -91,6 +91,36 @@ export function isArrayOfClass(object, expectedClass) {
...
 }
 
+export function isInterface(object, expectedProperties) {
+  const parameter = Object.keys(object).pop();
+  const message = `The '${parameter}' parameter should be an object that has
+    '${expectedProperties}' properties.`;
+
+  for (let property of expectedProperties) {
+    if (!(property in object[parameter])) {
+      throwError(message);
```
Comment (gauntface) [inline lib/assert.js]: Would it be worth adding the missing parameter to the message?

**Comment by gauntface** [comment]
In the early versions we originally exposed it, but developer feedback was that it caused confusion have it exposed so closely to router.

File: packages/workbox-sw/src/lib/router.js [hunk 6]
```diff
@@ -97,9 +99,9 @@ class Router extends SWRoutingRouter {
       if (capture.length === 0) {
         throw ErrorFactory.createError('empty-express-string');
       }
-      route = new ExpressRoute({path: capture, handler});
+      route = new ExpressRoute({path: capture, handler, method});
```
Comment (gauntface) [inline packages/workbox-sw/src/lib/router.js]: why did you add the method?

**Comment by gauntface** [comment]
Discussed with @jeffposnick , we can move the capture argument to accept a function, this will serve as the 'match' input for a Route that will be constructed behinds the scenes.

Sorry for this wrinkle.There is a largely re-think we should probably have about how we expose these values.

**Comment by mikefowler** [comment]
@gauntface, to clarify, your proposal is to allow for the following? 

```js
workboxSW.router.registerRoute(
  
  // Capture
  ({ url, event }) => {
    return true;
  },
  
  // Handler
  ({ event }) => {
    return caches.match(event.request);
  },

);

**Comment by jeffposnick** [comment]
That's correct, @mikefowler. I'm going to create a fresh PR that takes that approach.
